### DISCUSSION TITLE
How to debug aab with Play Asset Delivery

### DISCUSSION THREAD

**Discussion by jimmymjing** [airsdk/Adobe-Runtime-Support#1477] [Q&A] [answered]
Hello, I've packaged the aab for Google Play and I want to debug Play Asset Delivery part of the code. How am I going to debug it in IntelliJ? Thanks so much!

**Comment by ajwfrost** [comment, marked as answer]
There are a few suggestions from Android on how you can actually test play asset delivery:
https://developer.android.com/guide/app-bundle/test
https://developer.android.com/guide/playcore/feature-delivery/on-demand#local-testing

If you build using aab-debug then you should find when you start up the application on the device manually, it will try to connect to a debugger instance, so you can then kick off the 'remote debug' connection from the IDE.

thanks

**Reply by jimmymjing** [reply]
Thanks so much! I also noticed and tried to use `aab-debug` but didn't know how to use it...
btw, I'm using IntelliJ, and do you know how to kick off the remote debug connection from it? Do I do it in the GUI or by using command? (I usually just click the debug icon and the IDE will build -> compile -> and run the debugging session)
Thanks again!

**Reply by ajwfrost** [reply]
I think in IntelliJ you can configure a debug profile so that it doesn't do any build first, and so it debugs over the network .. 
Although I'm just trying it and it's complaining that there aren't any Android devices connected to the computer.. maybe it will also need to be connected via USB..

**Reply by ajwfrost** [reply]
... and actually it seems like with IntelliJ, it will always try to package and deploy the APK file (even if the project hasn't changed?!) - so this wouldn't then work...

At this point I'd be going onto using 'trace' commands and hooking it up to Scout, to find out what's going on. Not quite as good as debugging I'm afraid...

thanks

**Reply by jimmymjing** [reply]
Thanks for helping me out! I couldn't find a way to create a profile without do build... I have created the aab with `aab-debug` and launched it on the usb connected device, but the IntelliJ debugger couldn't get triggered.

**Reply by jimmymjing** [reply]
Hello again, may I ask what IDE do you use for debugging? As `aab-debug` is available, devs must be able to utilize it somehow. Thanks!

**Reply by marchbold** [reply]
@jimmymjing Have a look at the "Flash Remote" run configuration, this will launch intellij into debugging mode without doing a build.

**Reply by jimmymjing** [reply]
Glad to hear from you! Thanks so much and I will check it right away! BTW, The PlayAssetDeliveryEvent Class is missing. I posted an issue in that repo. Please take a look.

**Reply by ajwfrost** [reply]
@jimmymjing - I can't see an issue raised for this in the repo but I just checked the .ane file that I can download from it, and expanded this to grab the library.swf. This does include the PlayAssetDeliveryEvent:
```
ABC Block 0 Classes
 - com.harman.extension::PlayAssetDeliveryEvent$
```
so can I check where/how you're looking?

**Reply by jimmymjing** [reply]
It just got fixed about 4 hours ago. Thanks! I should have mentioned it here as well.
Here's the original issue page: https://github.com/airsdk/ANE-PlayAssetDelivery/issues/2

**Reply by ajwfrost** [reply]
Ah, okay thanks - I have belatedly added myself as a watcher on that repo ...
Thanks @marchbold for the inputs/updates! :-)

**Comment by jimmymjing** [comment]
I'm trying to figure out a way to integrate PAD into Starling's AssetManager's Class. It can be quite hard to do this without a proper debugger... So the codebase could work for both iOS and Android.
I've mentioned this here as well. https://github.com/airsdk/ANE-PlayAssetDelivery/issues/1

**Comment by TMTHawkeye** [comment]
There are commands to be typed on CMD, inorder to test debug.aab.

Following are the steps:

1- Set environment variable: Go to windows search, and type environment variables, a dialog will appear, select path from system variables, click edit, add jdk path and click ok for all dialogs.

2- Now generate debug bundle. Locate the path. Add bundle tools.jar file. You can fetch it from internet. Type cmd on path of this folder. 

3- use this command on cmd:

java -jar bundletools.jar build-apks --local-testing --bundle=app-debug.aab --output=apk.apks

4- then use this command: 

java -jar bundletools.jar install-apks --apks=apk.apks --adb="C:\Users\ApptrickTechnologies\AppData\Local\Android\Sdk\platform-tools\adb.exe"

Make sure your device is connected through adb to the system. 

This will generate the apk and run it in device. That is how dynamic assets will be fetched in debug mode
